PopPUNK/utils.py

```diff
@@ -497,3 +494,162 @@ def isolateNameToLabel(names):
     # want to remove certain characters
     labels = [name.split('/')[-1].split('.')[0] for name in names]
     return labels
+
+
+def sketchlib_assembly_qc(assemblyList, prefix, klist, estimated_length,
+                 qc_filter, retain_failures, length_sigma, lower_length, upper_length, prop_n,
+                 upper_n, strand_preserved, threads):
+    """Calculates random match probability based on means of genomes
+    in assemblyList, and looks for length outliers.
+
+    Args:
+        assemblyList (str)
+            File with locations of assembly files to be sketched
+        prefix (str)
+            Prefix of output files
+        klist (list)
+            List of k-mer sizes to sketch
+        estimated_length (int)
+            Estimated length of genome, if not calculated from data
+        qc_filter (string)
+            Behaviour on identifying sequences failing QC
+        retain_failures (bool)
+            Keep sketches of genomes that fail QC separate from main
+            database
+        length_sigma (int)
+            Number of SDs of length distribution beyond which sequences
+            are excluded
+        lower_length (int)
+            Threshold length below which sequences are excluded
+        upper_length (int)
+            Threshold length above which sequences are excluded
+        prop_n (float)
+            Proportion of ambiguous bases above which sequences are excluded
+        upper_n (int)
+            Number of ambiguous bases above which sequences are excluded
+        strand_preserved (bool)
+            Ignore reverse complement k-mers (default = False)
+        threads (int)
+            Number of threads to use in parallelisation
+
+    Returns:
+        genome_length (int)
+            Average length of assemblies
+    """
+    # Genome length needed to calculate prob of random matches
+    if estimated_length is not None:
+        mean_genome_length = estimated_length # assume 2 Mb in the absence of other information
+
+    # open databases
+    db_name = prefix + '/' + os.path.basename(prefix) + '.h5'
+    filtered_db_name = prefix + '/' + 'filtered.' + os.path.basename(prefix) + '.h5'
+    failed_db_name = prefix + '/' + 'failed.' + os.path.basename(prefix) + '.h5'
+    hdf_in = h5py.File(db_name, 'r')
+    # new database file if pruning
+    if qc_filter == 'prune':
+        hdf_out = h5py.File(filtered_db_name, 'w')
+    # retain sketches of failed samples
+    if retain_failures:
+        hdf_fail = h5py.File(failed_db_name, 'w')
+    
+    # process data structures
+    read_grp = hdf_in['sketches']
+    if qc_filter == 'prune':
+        out_grp = hdf_out.create_group('sketches')
+    if retain_failures:
+        fail_grp = hdf_fail.create_group('sketches')
+    seq_length = {}
+    seq_ambiguous = {}
+    seq_excluded = {}
+    removed = []
+    retained = []
+    
+    # iterate through sketches
+    for dataset in read_grp:
+        # test thresholds
+        remove = False
```

Comment:
Not needed

Reply:
Done in e601f73.

Reply:
It *is* as capricious as I thought! Tests failed, `remove = False` added back in bcc7956.